enzy_htp/structure/structure_constraint/api.py

```diff
@@ -1,43 +1,553 @@
+    + AngleConstraint: Constraint that specifies an Angle.
+    + DihedralConstraint: Constraint that specifies a dihedral.
+    + ResiduePairConstraint: Specifies a constraint set between two residues in a style similar to Rosetta EnzDes.
 
 Author: Qianzhen (QZ) Shao, <shaoqz@icloud.com>
+Author: Chris Jurich <chris.jurich@vanderbilt.edu>
 Date: 2022-10-28
 """
+from __future__ import annotations
 import copy
-from typing import List, Tuple
+from copy import deepcopy
+from typing import List, Tuple, Dict, Any
+
+import numpy as np
 
 from enzy_htp.core.logger import _LOGGER
 from ..structure import Structure, Solvent, Chain, Residue, Atom
 
-class StructureConstraint:
-    """This class describe a constrain of a structure (defined by the topology)
-    Constructors:
-        from_keyword()
+from abc import ABC, abstractmethod
+
+
+class StructureConstraint(ABC):
+    """Abstract base class defining the API for a constraint. Each primitive StructureConstraint defines exactly one 
+    type of interaction. StructureConstraints are meant to define flexible, non-package specific relationships that
+    can be translated in between different software packages.
+
     Attributes:
-        topology: Structure
-        freeze_atoms: List[Atom]
-        geom_constrain: List[Tuple[List[Atom], float]]
-
-    Derived properties:
-        structures: List[Structure]"""
-
-    def __init__(self, topology: Structure,
-                 freeze_atoms: List[Atom], geom_constrain: List[Tuple[List[Atom], float]]):
-        self._topology = topology
-        self._freeze_atoms = freeze_atoms
-        self._geom_constrain = geom_constrain
-        # init for check
-        self._is_backbone_freeze = None
-
-    def is_backbone_freeze(self) -> bool:
-        """this method saves effort when parsing to common Amber constrain"""
-        if self._is_backbone_freeze is None:
-            return set(self._topology.backbone_atoms()) == set(self._freeze_atoms)
+        atoms_: A List[Atom] containing all the atoms in the constraint.
+        target_value_ : The float() target value for the constraint.
+        params_: A Dict[str,float] with arbitrary parameters.
+    """
+
+    def __init__(self,
+                atoms:List[Atom],
+                target_value:float,
+                params:Dict[str, Any]
+                ):
+        """Ctor that exits if the supplied number of atoms is NOT correct."""
+        
+        self.atoms_ = atoms 
+        self.target_value_ = target_value 
+        self.params_ = deepcopy( params ) 
+
+        self['calc_method'] = params.get('calc_method', 'rosetta')
+
+        if not self.correct_num_atoms():
+            _LOGGER.error("Incorrect number of atoms supplied! Exiting...")
+            exit( 1 )
+
+
+    #TODO(CJ): add function that checks if topology and constraints are compatible
+    #TODO(CJ): will need to make a version of this that actually works for the ResiduePairConstraint
```

Comment:
probably a typo here? Not it is not finished yet.

Reply:
yes. not finished yet

Reply:
cool